Allow promoting and singling of visible type applications in data constructor patterns

GHC 9.2 introduces the ability to have [visible type applications in data constructor patterns](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/2464) like so:

```hs
blah :: Maybe a -> [a]
blah (Just @a x)  = [x :: a]
blah (Nothing @a) = [] :: [a]
```

Can we promote and single such a definition in `singletons`? While supporting visible type applications in their full generality in `singletons` is challenging (see #378), visible type applications in data constructor patterns are a relatively self-contained, well behaved part of `TypeApplications`. This is because `singletons` can reliably guarantee that promoted and singled data constructors will have the same behavior _vis-à-vis_ visible type applications as their term-level counterparts (see #408), which means that as long as we restrict ourselves to data constructors, the `TypeApplications`-related issues mentioned in #378 shouldn't arise.

Here is a first attempt at implementing this:

<details>

```diff
diff --git a/singletons-th/src/Data/Singletons/TH/Promote.hs b/singletons-th/src/Data/Singletons/TH/Promote.hs
index 373c35e..cb889b4 100644
--- a/singletons-th/src/Data/Singletons/TH/Promote.hs
+++ b/singletons-th/src/Data/Singletons/TH/Promote.hs
@@ -1051,12 +1051,11 @@ promotePat (DVarP name) = do
   tell $ PromDPatInfos [(name, tyName)] OSet.empty
   return (DVarT tyName, ADVarP name)
 promotePat (DConP name tys pats) = do
-  unless (null tys) $
-    qReportWarning "Visible type applications in patterns are ignored by `singletons-th`."
   opts <- getOptions
+  kis <- traverse promoteType tys
   (types, pats') <- mapAndUnzipM promotePat pats
   let name' = promotedDataTypeOrConName opts name
-  return (foldType (DConT name') types, ADConP name pats')
+  return (foldType (foldl DAppKindT (DConT name') kis) types, ADConP name kis pats')
 promotePat (DTildeP pat) = do
   qReportWarning "Lazy pattern converted into regular pattern in promotion"
   second ADTildeP <$> promotePat pat
diff --git a/singletons-th/src/Data/Singletons/TH/Single.hs b/singletons-th/src/Data/Singletons/TH/Single.hs
index 6096539..e8564b8 100644
--- a/singletons-th/src/Data/Singletons/TH/Single.hs
+++ b/singletons-th/src/Data/Singletons/TH/Single.hs
@@ -790,9 +790,9 @@ singPat var_proms = go
                   Just tyname -> return tyname
       pure $ DVarP (singledValueName opts name)
                `DSigP` (singFamily `DAppT` DVarT tyname)
-    go (ADConP name pats) = do
+    go (ADConP name tys pats) = do
       opts <- getOptions
-      DConP (singledDataConName opts name) [] <$> mapM go pats
+      DConP (singledDataConName opts name) tys <$> mapM go pats
     go (ADTildeP pat) = do
       qReportWarning
         "Lazy pattern converted into regular pattern during singleton generation."
diff --git a/singletons-th/src/Data/Singletons/TH/Syntax.hs b/singletons-th/src/Data/Singletons/TH/Syntax.hs
index 523341e..0297407 100644
--- a/singletons-th/src/Data/Singletons/TH/Syntax.hs
+++ b/singletons-th/src/Data/Singletons/TH/Syntax.hs
@@ -114,7 +114,7 @@ data ADExp = ADVarE Name
 -- A DPat with a pattern-signature node annotated with its type-level equivalent
 data ADPat = ADLitP Lit
            | ADVarP Name
-           | ADConP Name [ADPat]
+           | ADConP Name [DType] [ADPat]
            | ADTildeP ADPat
            | ADBangP ADPat
            | ADSigP DType -- The promoted pattern. Will not contain any wildcards,
```
</details>

This suffices to allow the `blah` function above to be promoted and singled. There's one remaining issue left to fix: wildcard patterns. If you have this program:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -ddump-splices #-}
module VTAPats where

import Data.Singletons.Base.TH
import Prelude.Singletons

$(singletons [d|
  argh :: Maybe () -> ()
  argh (Nothing @_) = ()
  argh (Just ())    = ()
  |])
```

Then it will fail like so:

```
[1 of 1] Compiling VTAPats          ( Bug.hs, interpreted )

Bug.hs:14:2: error: Q monad failure
   |
14 | $(singletons [d|
   |  ^^^^^^^^^^^^^^^...

Bug.hs:14:2: error:
    Illegal Haskell construct encountered:
headed by: DWildCardT
applied to: []
   |
14 | $(singletons [d|
   |  ^^^^^^^^^^^^^^^...
```

See #480.

--- a/README.md
+++ b/README.md
@@ -811,6 +811,8 @@ The following constructs are partially supported:
 * signatures (e.g., `(x :: Maybe a)`) in patterns
 * functional dependencies
 * type families
+* `TypeApplications`
+* wildcard types
 
 See the following sections for more details.
 
@@ -922,6 +924,147 @@ Promoting functions with types that contain type families is likely to fail due 
 [GHC#12564](https://gitlab.haskell.org/ghc/ghc/issues/12564).
 Note that promoting type family _declarations_ is fine
 (and often desired, since that produces defunctionalization symbols for them).
+
+### `TypeApplications`
+
+Currently, `singletons-th` will only promote or single code that uses
+`TypeApplications` syntax in one specific situation: when type applications
+are used in data constructor patterns, such as in the following example:
+
+```haskell
+foo :: Maybe a -> [a]
+foo (Just @a x)  = [x :: a]
+foo (Nothing @a) = [] :: [a]
+```
+
+For all other forms of `TypeApplications`, `singletons-th` will simply drop any
+visible type applications. For example, `id @Bool True` will be promoted to
+`Id True` and singled to `sId STrue`. See
+[#378](https://github.com/goldfirere/singletons/issues/378) for a discussion
+of how `singletons-th` may support `TypeApplications` in more places in the
+future.
+
+On the other hand, `singletons-th` does make an effort to preserve the order of
+type variables when promoting and singling certain constructors. These include:
+
+* Kind signatures of promoted top-level functions
+* Type signatures of singled top-level functions
+* Kind signatures of singled data type declarations
+* Type signatures of singled data constructors
+* Kind signatures of singled class declarations
+* Type signatures of singled class methods
+
+For example, consider this type signature:
+
+```haskell
+const2 :: forall b a. a -> b -> a
+```
+
+The promoted version of `const` will have the following kind signature:
+
+```haskell
+type Const2 :: forall b a. a -> b -> a
+```
+
+The singled version of `const2` will have the following type signature:
+
+```haskell
+sConst2 :: forall b a (x :: a) (y :: a). Sing x -> Sing y -> Sing (Const x y)
+```
+
+Therefore, writing `const2 @T1 @T2` works just as well as writing
+`Const2 @T1 @T2` or `sConst2 @T1 @T2`, since the signatures for `const2`, `Const2`,
+and `sConst2` all begin with `forall b a.`, in that order. Again, it is worth
+emphasizing that the TH machinery does not support promoting or singling
+`const2 @T1 @T2` directly, but you can write the type applications by hand if
+you so choose.
+
+`singletons-th` also has limited support for preserving the order of type variables
+for the following constructs:
+
+* Kind signatures of defunctionalization symbols.
+  The order of type variables is only guaranteed to be preserved if:
+
+  1. The thing being defunctionalized has a standalone type (or kind)
+     signature.
+  2. The type (or kind) signature of the thing being defunctionalized is
+     a vanilla type. (See the "Rank-n types" section above for what "vanilla"
+     means.)
+
+  If either of these conditions do not hold, `singletons-th` will fall back to
+  a slightly different approach to generating defunctionalization symbols that
+  does *not* guarantee the order of type variables. As an example, consider the
+  following example:
+
+  ```haskell
+  data T (x :: a) :: forall b. b -> Type
+  $(genDefunSymbols [''T])
+  ```
+
+  The kind of `T` is `forall a. a -> forall b. b -> Type`, which is not
+  vanilla. Currently, `singletons-th` will generate the following
+  defunctionalization symbols for `T`:
+
+  ```haskell
+  data TSym0 :: a ~> b ~> Type
+  data TSym1 (x :: a) :: b ~> Type
+  ```
+
+  In both symbols, the kind starts with `forall a b.` rather than quantifying
+  the `b` after the visible argument of kind `a`. These symbols can still be
+  useful even with this flaw, so `singletons-th` permits generating them
+  regardless. Be aware of this drawback if you try doing something similar
+  yourself!
+
+* Kind signatures of promoted class methods.
+  The order of type variables will often "just work" by happy coincidence, but
+  there are some situations where this does not happen. Consider the following
+  class:
+
+  ```haskell
+  class C (b :: Type) where
+    m :: forall a. a -> b -> a
+  ```
+
+  The full type of `m` is `forall b. C b => forall a. a -> b -> a`, which binds
+  `b` before `a`. This order is preserved when singling `m`, but *not* when
+  promoting `m`. This is because the `C` class is promoted as follows:
+
+  ```haskell
+  class PC (b :: Type) where
+    type M (x :: a) (y :: b) :: a
+  ```
+
+  Due to the way GHC kind-checks associated type families, the kind of `M` is
+  `forall a b. a -> b -> a`, which binds `b` *after* `a`. Moreover, the
+  `StandaloneKindSignatures` extension does not provide a way to explicitly
+  declare the full kind of an associated type family, so this limitation is
+  not easy to work around.
+
+  The defunctionalization symbols for `M` will also follow a similar
+  order of type variables:
+
+  ```haskell
+  type MSym0 :: forall a b. a ~> b ~> a
+  type MSym1 :: forall a b. a -> b ~> a
+  ```
+
+### Wildcard types
+
+Currently, `singletons-th` can only handle promoting or singling wildcard types
+if they appear within type applications in data constructor patterns, such as
+in the following example:
+
+```hs
+bar :: Maybe () -> Bool
+bar (Nothing @_) = False
+bar (Just ())    = True
+```
+
+`singletons-th` will error if trying to promote or single a wildcard type in
+any other context. Ultimately, this is due to a GHC restriction, as GHC itself
+will forbid using wildcards in most kind-level contexts. For example, GHC will
+permit `f :: _` but reject `type F :: _`.
 
 ## Support for promotion, but not singling
 
@@ -1316,120 +1459,6 @@ mechanism will not promote `TypeRep` to `*`.
 2) The module `Data.Singletons.TH.CustomStar` (from `singletons-th`) allows the programmer to define a subset
 of types with which to work. See the Haddock documentation for the function
 `singletonStar` for more info.
-
-### `TypeApplications`
-
-`singletons-th` currently cannot handle promoting or singling code that uses
-`TypeApplications` syntax, so the Template Haskell machinery will simply drop
-any visible type applications. For example, `id @Bool True` will be promoted to
-`Id True` and singled to `sId STrue`. See
-[#378](https://github.com/goldfirere/singletons/issues/378) for a discussion
-of how `singletons-th` may support `TypeApplications` in the future.
-
-On the other hand, `singletons-th` does make an effort to preserve the order of
-type variables when promoting and singling certain constructors. These include:
-
-* Kind signatures of promoted top-level functions
-* Type signatures of singled top-level functions
-* Kind signatures of singled data type declarations
-* Type signatures of singled data constructors
-* Kind signatures of singled class declarations
-* Type signatures of singled class methods
-
-For example, consider this type signature:
-
-```haskell
-const2 :: forall b a. a -> b -> a
-```
-
-The promoted version of `const` will have the following kind signature:
-
-```haskell
-type Const2 :: forall b a. a -> b -> a
-```
-
-The singled version of `const2` will have the following type signature:
-
-```haskell
-sConst2 :: forall b a (x :: a) (y :: a). Sing x -> Sing y -> Sing (Const x y)
-```
-
-Therefore, writing `const2 @T1 @T2` works just as well as writing
-`Const2 @T1 @T2` or `sConst2 @T1 @T2`, since the signatures for `const2`, `Const2`,
-and `sConst2` all begin with `forall b a.`, in that order. Again, it is worth
-emphasizing that the TH machinery does not support promoting or singling
-`const2 @T1 @T2` directly, but you can write the type applications by hand if
-you so choose.
-
-`singletons-th` also has limited support for preserving the order of type variables
-for the following constructs:
-
-* Kind signatures of defunctionalization symbols.
-  The order of type variables is only guaranteed to be preserved if:
-
-  1. The thing being defunctionalized has a standalone type (or kind)
-     signature.
-  2. The type (or kind) signature of the thing being defunctionalized is
-     a vanilla type. (See the "Rank-n types" section above for what "vanilla"
-     means.)
-
-  If either of these conditions do not hold, `singletons-th` will fall back to
-  a slightly different approach to generating defunctionalization symbols that
-  does *not* guarantee the order of type variables. As an example, consider the
-  following example:
-
-  ```haskell
-  data T (x :: a) :: forall b. b -> Type
-  $(genDefunSymbols [''T])
-  ```
-
-  The kind of `T` is `forall a. a -> forall b. b -> Type`, which is not
-  vanilla. Currently, `singletons-th` will generate the following
-  defunctionalization symbols for `T`:
-
-  ```haskell
-  data TSym0 :: a ~> b ~> Type
-  data TSym1 (x :: a) :: b ~> Type
-  ```
-
-  In both symbols, the kind starts with `forall a b.` rather than quantifying
-  the `b` after the visible argument of kind `a`. These symbols can still be
-  useful even with this flaw, so `singletons-th` permits generating them
-  regardless. Be aware of this drawback if you try doing something similar
-  yourself!
-
-* Kind signatures of promoted class methods.
-  The order of type variables will often "just work" by happy coincidence, but
-  there are some situations where this does not happen. Consider the following
-  class:
-
-  ```haskell
-  class C (b :: Type) where
-    m :: forall a. a -> b -> a
-  ```
-
-  The full type of `m` is `forall b. C b => forall a. a -> b -> a`, which binds
-  `b` before `a`. This order is preserved when singling `m`, but *not* when
-  promoting `m`. This is because the `C` class is promoted as follows:
-
-  ```haskell
-  class PC (b :: Type) where
-    type M (x :: a) (y :: b) :: a
-  ```
-
-  Due to the way GHC kind-checks associated type families, the kind of `M` is
-  `forall a b. a -> b -> a`, which binds `b` *after* `a`. Moreover, the
-  `StandaloneKindSignatures` extension does not provide a way to explicitly
-  declare the full kind of an associated type family, so this limitation is
-  not easy to work around.
-
-  The defunctionalization symbols for `M` will also follow a similar
-  order of type variables:
-
-  ```haskell
-  type MSym0 :: forall a b. a ~> b ~> a
-  type MSym1 :: forall a b. a -> b ~> a
-  ```
 
 ### Irrefutable patterns
 

--- a/singletons-base/tests/SingletonsBaseTestSuite.hs
+++ b/singletons-base/tests/SingletonsBaseTestSuite.hs
@@ -138,6 +138,7 @@ tests =
     , compileAndDumpStdTest "T453"
     , compileAndDumpStdTest "NegativeLiterals"
     , compileAndDumpStdTest "T470"
+    , compileAndDumpStdTest "T480"
     , compileAndDumpStdTest "T487"
     , compileAndDumpStdTest "T492"
     , compileAndDumpStdTest "Natural"

--- a/singletons-base/tests/SingletonsBaseTestSuite.hs
+++ b/singletons-base/tests/SingletonsBaseTestSuite.hs
@@ -140,6 +140,7 @@ tests =
     , compileAndDumpStdTest "T470"
     , compileAndDumpStdTest "T480"
     , compileAndDumpStdTest "T487"
+    , compileAndDumpStdTest "T489"
     , compileAndDumpStdTest "T492"
     , compileAndDumpStdTest "Natural"
     ],

--- a/singletons-base/tests/compile-and-dump/Singletons/T480.golden
+++ b/singletons-base/tests/compile-and-dump/Singletons/T480.golden
@@ -6,6 +6,7 @@ Singletons/T480.hs:0:0: error: Q monad failure
 
 Singletons/T480.hs:0:0: error:
     `singletons-th` does not support wildcard types
+	unless they appear in visible type patterns of data constructors
 	In the type: _
 
   |

--- a/singletons-base/tests/compile-and-dump/Singletons/T480.golden
+++ b/singletons-base/tests/compile-and-dump/Singletons/T480.golden
@@ -1,0 +1,13 @@
+
+Singletons/T480.hs:0:0: error: Q monad failure
+  |
+5 | $(singletons
+  |  ^^^^^^^^^^^...
+
+Singletons/T480.hs:0:0: error:
+    `singletons-th` does not support wildcard types
+	In the type: _
+
+  |
+5 | $(singletons
+  |  ^^^^^^^^^^^...

--- a/singletons-base/tests/compile-and-dump/Singletons/T480.hs
+++ b/singletons-base/tests/compile-and-dump/Singletons/T480.hs
@@ -1,0 +1,8 @@
+module T480 where
+
+import Data.Singletons.TH
+
+$(singletons
+  [d| f :: _ -> _
+      f x = x
+  |])

--- a/singletons-base/tests/compile-and-dump/Singletons/T489.golden
+++ b/singletons-base/tests/compile-and-dump/Singletons/T489.golden
@@ -1,0 +1,60 @@
+Singletons/T489.hs:(0,0)-(0,0): Splicing declarations
+    singletons
+      [d| blah :: Maybe a -> [a]
+          blah (Just @a x) = [x :: a]
+          blah (Nothing @a) = [] :: [a]
+          flurmp :: Maybe () -> ()
+          flurmp (Nothing @_) = ()
+          flurmp (Just ()) = () |]
+  ======>
+    blah :: Maybe a -> [a]
+    blah (Just @a x) = [x :: a]
+    blah (Nothing @a) = [] :: [a]
+    flurmp :: Maybe () -> ()
+    flurmp (Nothing @_) = ()
+    flurmp (Just ()) = ()
+    type FlurmpSym0 :: (~>) (Maybe ()) ()
+    data FlurmpSym0 :: (~>) (Maybe ()) ()
+      where
+        FlurmpSym0KindInference :: SameKind (Apply FlurmpSym0 arg) (FlurmpSym1 arg) =>
+                                   FlurmpSym0 a0123456789876543210
+    type instance Apply FlurmpSym0 a0123456789876543210 = Flurmp a0123456789876543210
+    instance SuppressUnusedWarnings FlurmpSym0 where
+      suppressUnusedWarnings = snd (((,) FlurmpSym0KindInference) ())
+    type FlurmpSym1 :: Maybe () -> ()
+    type family FlurmpSym1 (a0123456789876543210 :: Maybe ()) :: () where
+      FlurmpSym1 a0123456789876543210 = Flurmp a0123456789876543210
+    type BlahSym0 :: (~>) (Maybe a) [a]
+    data BlahSym0 :: (~>) (Maybe a) [a]
+      where
+        BlahSym0KindInference :: SameKind (Apply BlahSym0 arg) (BlahSym1 arg) =>
+                                 BlahSym0 a0123456789876543210
+    type instance Apply BlahSym0 a0123456789876543210 = Blah a0123456789876543210
+    instance SuppressUnusedWarnings BlahSym0 where
+      suppressUnusedWarnings = snd (((,) BlahSym0KindInference) ())
+    type BlahSym1 :: Maybe a -> [a]
+    type family BlahSym1 (a0123456789876543210 :: Maybe a) :: [a] where
+      BlahSym1 a0123456789876543210 = Blah a0123456789876543210
+    type Flurmp :: Maybe () -> ()
+    type family Flurmp (a :: Maybe ()) :: () where
+      Flurmp ('Nothing @_) = Tuple0Sym0
+      Flurmp ('Just '()) = Tuple0Sym0
+    type Blah :: Maybe a -> [a]
+    type family Blah (a :: Maybe a) :: [a] where
+      Blah ('Just @a x) = Apply (Apply (:@#@$) (x :: a)) NilSym0
+      Blah ('Nothing @a) = NilSym0 :: [a]
+    sFlurmp ::
+      forall (t :: Maybe ()). Sing t -> Sing (Apply FlurmpSym0 t :: ())
+    sBlah ::
+      forall a (t :: Maybe a). Sing t -> Sing (Apply BlahSym0 t :: [a])
+    sFlurmp (SNothing @_) = STuple0
+    sFlurmp (SJust STuple0) = STuple0
+    sBlah (SJust @a (sX :: Sing x))
+      = (applySing
+           ((applySing ((singFun2 @(:@#@$)) SCons)) (sX :: Sing (x :: a))))
+          SNil
+    sBlah (SNothing @a) = SNil :: Sing (NilSym0 :: [a])
+    instance SingI (FlurmpSym0 :: (~>) (Maybe ()) ()) where
+      sing = (singFun1 @FlurmpSym0) sFlurmp
+    instance SingI (BlahSym0 :: (~>) (Maybe a) [a]) where
+      sing = (singFun1 @BlahSym0) sBlah

--- a/singletons-base/tests/compile-and-dump/Singletons/T489.hs
+++ b/singletons-base/tests/compile-and-dump/Singletons/T489.hs
@@ -1,0 +1,14 @@
+module T489 where
+
+import Data.Singletons.Base.TH
+import Prelude.Singletons
+
+$(singletons [d|
+  blah :: Maybe a -> [a]
+  blah (Just @a x)  = [x :: a]
+  blah (Nothing @a) = [] :: [a]
+
+  flurmp :: Maybe () -> ()
+  flurmp (Nothing @_) = ()
+  flurmp (Just ())    = ()
+  |])

--- a/singletons-th/CHANGES.md
+++ b/singletons-th/CHANGES.md
@@ -4,6 +4,7 @@ Changelog for the `singletons-th` project
 next [????.??.??]
 -----------------
 * Require building with GHC 9.2.
+* Allow promoting and singling type applications in data constructor patterns.
 * Make the Template Haskell machinery generate `SingI1` and `SingI2` instances
   when possible.
 * Make `genDefunSymbols` and related functions less likely to trigger

--- a/singletons-th/src/Data/Singletons/TH/Promote.hs
+++ b/singletons-th/src/Data/Singletons/TH/Promote.hs
@@ -1051,12 +1051,17 @@ promotePat (DVarP name) = do
   tell $ PromDPatInfos [(name, tyName)] OSet.empty
   return (DVarT tyName, ADVarP name)
 promotePat (DConP name tys pats) = do
-  unless (null tys) $
-    qReportWarning "Visible type applications in patterns are ignored by `singletons-th`."
   opts <- getOptions
+  kis <- traverse (promoteType_options conOptions) tys
   (types, pats') <- mapAndUnzipM promotePat pats
   let name' = promotedDataTypeOrConName opts name
-  return (foldType (DConT name') types, ADConP name pats')
+  return (foldType (foldl DAppKindT (DConT name') kis) types, ADConP name kis pats')
+  where
+    -- Currently, visible type patterns of data constructors are the one place
+    -- in `singletons-th` where it makes sense to promote wildcard types, as it
+    -- will produce code that GHC will accept.
+    conOptions :: PromoteTypeOptions
+    conOptions = defaultPromoteTypeOptions{ptoAllowWildcards = True}
 promotePat (DTildeP pat) = do
   qReportWarning "Lazy pattern converted into regular pattern in promotion"
   second ADTildeP <$> promotePat pat

--- a/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
+++ b/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
@@ -9,26 +9,51 @@ This file implements promotion of types into kinds.
 {-# LANGUAGE ScopedTypeVariables #-}
 
 module Data.Singletons.TH.Promote.Type
-  ( promoteType, promoteType_NC
+  ( promoteType, promoteType_NC, promoteType_options
+  , PromoteTypeOptions(..), defaultPromoteTypeOptions
   , promoteTypeArg_NC, promoteUnraveled
   ) where
 
+import Control.Monad (when)
 import Language.Haskell.TH.Desugar
 import Data.Singletons.TH.Names
 import Data.Singletons.TH.Options
 import Data.Singletons.TH.Util
 
--- Promote a DType to the kind level.
+-- | Promote a 'DType' to the kind level and invoke 'checkVanillaDType'.
+-- See @Note [Vanilla-type validity checking during promotion]@.
 promoteType :: OptionsMonad m => DType -> m DKind
-promoteType ty = do
-  checkVanillaDType ty
-  promoteType_NC ty
+promoteType = promoteType_options defaultPromoteTypeOptions{ptoCheckVanilla = True}
 
--- Promote a DType to the kind level. This is suffixed with "_NC" because
--- we do not invoke checkVanillaDType here.
--- See [Vanilla-type validity checking during promotion].
+-- | Promote a 'DType' to the kind level. This is suffixed with \"_NC\" because
+-- we do not invoke 'checkVanillaDType' here.
+-- See @Note [Vanilla-type validity checking during promotion]@.
 promoteType_NC :: forall m. OptionsMonad m => DType -> m DKind
-promoteType_NC = go []
+promoteType_NC = promoteType_options defaultPromoteTypeOptions
+
+-- | Options for controlling how types are promoted at a fine granularity.
+newtype PromoteTypeOptions = PromoteTypeOptions
+  { ptoCheckVanilla :: Bool
+    -- ^ If 'True', invoke 'checkVanillaDType' on the argument type being
+    --   promoted. See @Note [Vanilla-type validity checking during promotion]@.
+  } deriving Show
+
+-- | The default 'PromoteTypeOptions':
+--
+-- * 'checkVanillaDType' is not invoked.
+defaultPromoteTypeOptions :: PromoteTypeOptions
+defaultPromoteTypeOptions = PromoteTypeOptions
+  { ptoCheckVanilla = False
+  }
+
+-- | Promote a 'DType' to the kind level. This is the workhorse for
+-- 'promoteType' and 'promoteType_NC'.
+promoteType_options :: forall m. OptionsMonad m => PromoteTypeOptions -> DType -> m DKind
+promoteType_options pto typ = do
+  -- See Note [Vanilla-type validity checking during promotion]
+  when (ptoCheckVanilla pto) $
+    checkVanillaDType typ
+  go [] typ
   where
     go :: [DTypeArg] -> DType -> m DKind
     go []       (DForallT tele ty) = do
@@ -77,7 +102,7 @@ promoteType_NC = go []
 
 -- | Promote a DTypeArg to the kind level. This is suffixed with "_NC" because
 -- we do not invoke checkVanillaDType here.
--- See [Vanilla-type validity checking during promotion].
+-- See @Note [Vanilla-type validity checking during promotion]@.
 promoteTypeArg_NC :: OptionsMonad m => DTypeArg -> m DTypeArg
 promoteTypeArg_NC (DTANormal t) = DTANormal <$> promoteType_NC t
 promoteTypeArg_NC ta@(DTyArg _) = pure ta -- Kinds are already promoted

--- a/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
+++ b/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
@@ -15,6 +15,7 @@ module Data.Singletons.TH.Promote.Type
   ) where
 
 import Control.Monad (when)
+import Language.Haskell.TH (pprint)
 import Language.Haskell.TH.Desugar
 import Data.Singletons.TH.Names
 import Data.Singletons.TH.Options
@@ -91,7 +92,10 @@ promoteType_options pto typ = do
     go args     ty@DArrowT = illegal args ty
     go []       ty@DLitT{} = pure ty
     go args     ty@DLitT{} = illegal args ty
-    go args     ty@DWildCardT{} = illegal args ty
+    go _args    DWildCardT{} = fail $ unlines
+      [ "`singletons-th` does not support wildcard types"
+      , "\tIn the type: " ++ pprint (sweeten typ)
+      ]
 
     illegal :: [DTypeArg] -> DType -> m a
     illegal args hd = fail $ unlines

--- a/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
+++ b/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
@@ -33,18 +33,27 @@ promoteType_NC :: forall m. OptionsMonad m => DType -> m DKind
 promoteType_NC = promoteType_options defaultPromoteTypeOptions
 
 -- | Options for controlling how types are promoted at a fine granularity.
-newtype PromoteTypeOptions = PromoteTypeOptions
+data PromoteTypeOptions = PromoteTypeOptions
   { ptoCheckVanilla :: Bool
     -- ^ If 'True', invoke 'checkVanillaDType' on the argument type being
     --   promoted. See @Note [Vanilla-type validity checking during promotion]@.
+  , ptoAllowWildcards :: Bool
+    -- ^ If 'True', allow promoting wildcard types. Otherwise, throw an error.
+    --   In most places, GHC disallows kind-level wildcard types, so rather
+    --   than promoting such wildcards and getting an error message from GHC
+    --   /post facto/, we can catch such wildcards early and give a more
+    --   descriptive error message instead.
   } deriving Show
 
 -- | The default 'PromoteTypeOptions':
 --
 -- * 'checkVanillaDType' is not invoked.
+--
+-- * Throw an error when attempting to promote a wildcard type.
 defaultPromoteTypeOptions :: PromoteTypeOptions
 defaultPromoteTypeOptions = PromoteTypeOptions
   { ptoCheckVanilla = False
+  , ptoAllowWildcards = False
   }
 
 -- | Promote a 'DType' to the kind level. This is the workhorse for
@@ -92,10 +101,15 @@ promoteType_options pto typ = do
     go args     ty@DArrowT = illegal args ty
     go []       ty@DLitT{} = pure ty
     go args     ty@DLitT{} = illegal args ty
-    go _args    DWildCardT{} = fail $ unlines
-      [ "`singletons-th` does not support wildcard types"
-      , "\tIn the type: " ++ pprint (sweeten typ)
-      ]
+    go args     ty@DWildCardT{}
+      | ptoAllowWildcards pto
+      = pure $ applyDType ty args
+      | otherwise
+      = fail $ unlines
+          [ "`singletons-th` does not support wildcard types"
+          , "\tunless they appear in visible type patterns of data constructors"
+          , "\tIn the type: " ++ pprint (sweeten typ)
+          ]
 
     illegal :: [DTypeArg] -> DType -> m a
     illegal args hd = fail $ unlines

--- a/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
+++ b/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
@@ -6,6 +6,8 @@ rae@cs.brynmawr.edu
 This file implements promotion of types into kinds.
 -}
 
+{-# LANGUAGE ScopedTypeVariables #-}
+
 module Data.Singletons.TH.Promote.Type
   ( promoteType, promoteType_NC
   , promoteTypeArg_NC, promoteUnraveled
@@ -25,18 +27,20 @@ promoteType ty = do
 -- Promote a DType to the kind level. This is suffixed with "_NC" because
 -- we do not invoke checkVanillaDType here.
 -- See [Vanilla-type validity checking during promotion].
-promoteType_NC :: OptionsMonad m => DType -> m DKind
+promoteType_NC :: forall m. OptionsMonad m => DType -> m DKind
 promoteType_NC = go []
   where
-    go :: OptionsMonad m => [DTypeArg] -> DType -> m DKind
+    go :: [DTypeArg] -> DType -> m DKind
     go []       (DForallT tele ty) = do
       ty' <- go [] ty
       pure $ DForallT tele ty'
+    go args     ty@DForallT{} = illegal args ty
     -- We don't need to worry about constraints: they are used to express
     -- static guarantees at runtime. But, because we don't need to do
     -- anything special to keep static guarantees at compile time, we don't
     -- need to promote them.
     go []       (DConstrainedT _cxt ty) = go [] ty
+    go args     ty@DConstrainedT{} = illegal args ty
     go args     (DAppT t1 t2) = do
       k2 <- go [] t2
       go (DTANormal k2 : args) t1
@@ -59,11 +63,17 @@ promoteType_NC = go []
       return $ applyDType (DConT (promotedDataTypeOrConName opts name)) args
     go [DTANormal k1, DTANormal k2] DArrowT
       = return $ DConT tyFunArrowName `DAppT` k1 `DAppT` k2
-    go _        ty@DLitT{} = pure ty
+    go args     ty@DArrowT = illegal args ty
+    go []       ty@DLitT{} = pure ty
+    go args     ty@DLitT{} = illegal args ty
+    go args     ty@DWildCardT{} = illegal args ty
 
-    go args     hd = fail $ "Illegal Haskell construct encountered:\n" ++
-                            "headed by: " ++ show hd ++ "\n" ++
-                            "applied to: " ++ show args
+    illegal :: [DTypeArg] -> DType -> m a
+    illegal args hd = fail $ unlines
+      [ "Illegal Haskell construct encountered:"
+      , "\theaded by: " ++ show hd
+      , "\tapplied to: " ++ show args
+      ]
 
 -- | Promote a DTypeArg to the kind level. This is suffixed with "_NC" because
 -- we do not invoke checkVanillaDType here.

--- a/singletons-th/src/Data/Singletons/TH/Single.hs
+++ b/singletons-th/src/Data/Singletons/TH/Single.hs
@@ -790,9 +790,9 @@ singPat var_proms = go
                   Just tyname -> return tyname
       pure $ DVarP (singledValueName opts name)
                `DSigP` (singFamily `DAppT` DVarT tyname)
-    go (ADConP name pats) = do
+    go (ADConP name tys pats) = do
       opts <- getOptions
-      DConP (singledDataConName opts name) [] <$> mapM go pats
+      DConP (singledDataConName opts name) tys <$> mapM go pats
     go (ADTildeP pat) = do
       qReportWarning
         "Lazy pattern converted into regular pattern during singleton generation."

--- a/singletons-th/src/Data/Singletons/TH/Syntax.hs
+++ b/singletons-th/src/Data/Singletons/TH/Syntax.hs
@@ -114,7 +114,7 @@ data ADExp = ADVarE Name
 -- A DPat with a pattern-signature node annotated with its type-level equivalent
 data ADPat = ADLitP Lit
            | ADVarP Name
-           | ADConP Name [ADPat]
+           | ADConP Name [DType] [ADPat]
            | ADTildeP ADPat
            | ADBangP ADPat
            | ADSigP DType -- The promoted pattern. Will not contain any wildcards,

